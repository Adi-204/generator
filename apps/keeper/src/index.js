import { registerSchema, validate, unregisterSchema } from '@hyperjump/json-schema/draft-07';
import { Parser, fromFile } from '@asyncapi/parser';

const parser = new Parser();

/**
* Validates a message payload against an array of operation message schemas.
* This method uses the Hyperjump JSON Schema validator (Draft-07) to check if
* the given message object conforms to at least one of the provided operation-level
* message schemas (as defined in the AsyncAPI document).
* @async
* @param {Object} message - The message payload to be validated.
* @param {Array<Object>} schemas - An array of JSON Schema objects representing operation message schemas.
* @returns {Promise<boolean>} - Resolves to true if the message is valid against any of the schemas, otherwise false.
*/
export async function validateMessage(message, schemas) {
  if (!Array.isArray(schemas)) {
    throw new Error(`Invalid "schemas" parameter: expected an array of JSON Schemas, but got ${typeof schemas}`);
  }
  if (!schemas || schemas.length === 0) {
    throw new Error('Empty "schemas" array: at least one JSON Schema must be provided for validation.');
  }
  if (message === null || message === undefined) {
    throw new Error(`Invalid "message" parameter: expected a non-null object to validate, but received ${message}`);
  }

  const dialectId = 'http://json-schema.org/draft-07/schema#';
  const schemaIds = [];

  try {
    // Register all schemas first
    for (let index = 0; index < schemas.length; index++) {
      const schemaId = `http://example.com/schema-${index}`;
      registerSchema(schemas[index], schemaId, dialectId);
      schemaIds.push(schemaId);
    }
    // Validate against each schema until first match
    for (const schemaId of schemaIds) {
      const result = await validate(schemaId, message);
      if (result.valid) {
        return true;
      }
    }
    return false;
  } catch (error) {
    console.error('Schema validation failed:', error);
    return false;
  } finally {
    // Clean up all registered schemas
    for (const schemaId of schemaIds) {
      try {
        unregisterSchema(schemaId);
      } catch (cleanupError) {
        console.error(`Failed to unregister schema ${schemaId}:`, cleanupError);
      }
    }
  }
}

/**
 * Validates a message payload against the schemas defined for a specific operation in an AsyncAPI document.
 * 
 * @async
 * @param {string} asyncapiFilepath - Path to the AsyncAPI document file.
 * @param {string} operationId - The operation ID to validate against.
 * @param {Object} message - The message payload to validate.
 * @returns {Promise<boolean>} - Resolves to true if the message is valid against any of the operation's message schemas.
 * @throws {Error} If the AsyncAPI document cannot be parsed, the operation is not found, or no message schemas exist.
 * 
 */
export async function validateByOperationId(asyncapiFilepath, operationId, message) {
  if (typeof asyncapiFilepath !== 'string' || !asyncapiFilepath.trim()) {
    throw new Error(`Invalid "asyncapiFilepath" parameter: must be a non-empty string, received ${asyncapiFilepath}`);
  }
  if (typeof operationId !== 'string' || !operationId.trim()) {
    throw new Error(`Invalid "operationId" parameter: must be a non-empty string, received ${operationId}`);
  }
  if (message === null || message === undefined) {
    throw new Error(`Invalid "message" parameter: expected a non-null object to validate, but received ${message}`);
  }

  let asyncapi;
  try {
    const parseResult = await fromFile(parser, asyncapiFilepath).parse();
    asyncapi = parseResult.document;
  } catch (error) {
    throw new Error(`Failed to parse AsyncAPI document: ${error.message}`);
  }

  const operation = asyncapi.operations().get(operationId);
  const messages = operation.messages().all();
  const messagePayloads = messages.map((message) => message.payload());
  const schemas = messagePayloads.map(payload => payload.json());

  return validateMessage(message, schemas);
}