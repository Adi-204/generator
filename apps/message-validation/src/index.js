import { registerSchema, validate, unregisterSchema } from '@hyperjump/json-schema/draft-07';

/**
* Validates a message payload against an array of operation message schemas.
* This method uses the Hyperjump JSON Schema validator (Draft-07) to check if
* the given message object conforms to at least one of the provided operation-level
* message schemas (as defined in the AsyncAPI document).
* @async
* @static
* @param {Object} message - The message payload to be validated.
* @param {Array<Object>} schemas - An array of JSON Schema objects representing operation message schemas.
* @returns {Promise<boolean>} - Resolves to true if the message is valid against any of the schemas, otherwise false.
*/
export async function validateMessage(message, schemas) {
  if (!Array.isArray(schemas)) {
    throw new Error(`Invalid "schemas" parameter: expected an array of JSON Schemas, but got ${typeof schemas}`);
  }
  if (!schemas || schemas.length === 0) {
    throw new Error('Empty "schemas" array: at least one JSON Schema must be provided for validation.');
  }
  if (message === null || message === undefined) {
    throw new Error(`Invalid "message" parameter: expected a non-null object to validate, but received ${message}`);
  }

  const dialectId = 'http://json-schema.org/draft-07/schema#';
  const schemaIds = [];

  try {
    // Register all schemas first
    for (let index = 0; index < schemas.length; index++) {
      const schemaId = `http://example.com/schema-${index}`;
      registerSchema(schemas[index], schemaId, dialectId);
      schemaIds.push(schemaId);
    }
    // Validate against each schema until first match
    for (const schemaId of schemaIds) {
      const result = await validate(schemaId, message);
      if (result.valid) {
        return true;
      }
    }
    return false;
  } catch (error) {
    console.error('Schema validation failed:', error);
    return false;
  } finally {
    // Clean up all registered schemas
    for (const schemaId of schemaIds) {
      try {
        unregisterSchema(schemaId);
      } catch (cleanupError) {
        console.warn(`Failed to unregister schema ${schemaId}:`, cleanupError);
      }
    }
  }
}