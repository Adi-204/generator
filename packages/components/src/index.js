export { Models } from './components/Models';
export { FileHeaderInfo } from './components/FileHeaderInfo';
export { FileDependencies } from './components/FileDependencies';