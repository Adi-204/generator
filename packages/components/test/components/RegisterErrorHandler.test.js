import { render } from '@asyncapi/generator-react-sdk';
import { RegisterErrorHandler } from '../../src/index';

describe('Testing of RegisterErrorHandler function', () => {
  test('render javascript websocket register error handler method', () => {
    const result = render(
      <RegisterErrorHandler language='javascript' methodParams={['handler']} />
    );
    const actual = result.trim();
    expect(actual).toMatchSnapshot();
  });

  test('render python websocket register error handler method', () => {
    const result = render(
      <RegisterErrorHandler 
        language='python'
        methodParams={['self', 'handler']}
        preExecutionCode='"""Register a callable as a error handler."""'
      />
    );
    const actual = result.trim();
    expect(actual).toMatchSnapshot();
  });

  test('render dart websocket register error handler method', () => {
    const result = render(
      <RegisterErrorHandler 
        language='dart'
        methodParams={['void Function(Object) handler']}
      />
    );
    const actual = result.trim();
    expect(actual).toMatchSnapshot();
  });

  test('render javascript method with custom name', () => {
    const result = render(
      <RegisterErrorHandler 
        language='javascript'
        methodName='addHandler'
        methodParams={['handler']}
      />
    );
    const actual = result.trim();
    expect(actual).toMatchSnapshot();
  });

  test('render dart method with parameters', () => {
    const result = render(
      <RegisterErrorHandler 
        language='dart'
        methodParams={['handler']}
      />
    );
    const actual = result.trim();
    expect(actual).toMatchSnapshot();
  });

  test('render python with pre and post execution code', () => {
    const result = render(
      <RegisterErrorHandler 
        language='python'
        methodParams={['self', 'handler']}
        preExecutionCode='# Pre-register operations'
        postExecutionCode='# Post-register cleanup'
      />
    );
    const actual = result.trim();
    expect(actual).toMatchSnapshot();
  });

  test('render dart with all the props', () => {
    const customMethodConfig = {
      returnType: 'int', 
      openingTag: '{', 
      closingTag: '}', 
      indentSize: 2
    };
    const result = render(
      <RegisterErrorHandler 
        language='dart'
        methodName='bindHandler'
        methodParams={['handler']}
        preExecutionCode='// Preparing to bind handler'
        postExecutionCode='// Handler bound successfully'
        customMethodConfig={customMethodConfig}
      />
    );
    const actual = result.trim();
    expect(actual).toMatchSnapshot();
  });
});
