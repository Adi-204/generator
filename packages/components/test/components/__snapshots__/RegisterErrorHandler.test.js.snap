// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing of RegisterErrorHandler function render dart method with parameters 1`] = `
"/// Method to register custom error handlers
  void registerErrorHandler(handler) {
    _errorHandlers.add(handler);
  }"
`;

exports[`Testing of RegisterErrorHandler function render dart websocket register error handler method 1`] = `
"/// Method to register custom error handlers
  void registerErrorHandler(void Function(Object) handler) {
    _errorHandlers.add(handler);
  }"
`;

exports[`Testing of RegisterErrorHandler function render dart with all the props 1`] = `
"/// Method to register custom error handlers
  int bindHandler(handler) {
    // Preparing to bind handler
    _errorHandlers.add(handler);
    // Handler bound successfully
  }"
`;

exports[`Testing of RegisterErrorHandler function render javascript method with custom name 1`] = `
"// Method to register custom error handlers
   addHandler(handler) {
    if (typeof handler === 'function') {
      this.errorHandlers.push(handler);
    } else {
      console.warn('Error handler must be a function');
    }
  }"
`;

exports[`Testing of RegisterErrorHandler function render javascript websocket register error handler method 1`] = `
"// Method to register custom error handlers
   registerErrorHandler(handler) {
    if (typeof handler === 'function') {
      this.errorHandlers.push(handler);
    } else {
      console.warn('Error handler must be a function');
    }
  }"
`;

exports[`Testing of RegisterErrorHandler function render python websocket register error handler method 1`] = `
"def registerErrorHandler(self, handler) :
    \\"\\"\\"Register a callable as a error handler.\\"\\"\\"
    if callable(handler):
      self.error_handlers.append(handler)
    else:
        print(\\"Error handler must be callable\\")"
`;

exports[`Testing of RegisterErrorHandler function render python with pre and post execution code 1`] = `
"def registerErrorHandler(self, handler) :
    # Pre-register operations
    if callable(handler):
      self.error_handlers.append(handler)
    else:
        print(\\"Error handler must be callable\\")
    # Post-register cleanup"
`;
